### PR TITLE
Add extra VIFs

### PR DESCRIPTION
**Pull request by jerith**
Sometimes a VM needs more than one network interface.

### CODE REVIEW

**Comment by jerith** [comment]
So far, this just does the backend provisioning work. The UI still needs to be updated.

**Comment by jerith** [comment]
Ready for review.

File: xenserver/tasks.py [hunk 1]
```diff
@@ -354,6 +358,18 @@ def _create_vm(session, vm, template, name, domain, ip, subnet, gateway, preseed
     if not network:
         raise NetworkError('Unable to locate VIF network')
 
+    extra_networks = []
+    for br in extra_network_bridges:
+        found = False
+        for net in session.xenapi.network.get_all():
```
Comment (JayH5) [inline xenserver/tasks.py]: Is there a reason the `network.get_all()` call is not done once before looping through the user-provided bridges?

Comment (jerith) [inline xenserver/tasks.py]: Yes, but it's not relevant to the current version of the code. I'll change that.

File: xenserver/tests/test_views.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            iso=iso)
+
+    def setup_project(self, name):
+        return Project.objects.create(name=name)
+
+    def setup_server(self, hostname):
+        zone = Zone.objects.create(name="foozone")
+        AddressPool.objects.create(
+            subnet="10.0.0.0/24", gateway="10.1.2.3", zone=zone, version=4)
+        mem = 64*1024*1204*1024
```
Comment (JayH5) [inline xenserver/tests/test_views.py]: Probably doesn't matter.. but maybe that should be 1024, not 1204

File: xenserver/tests/test_tasks.py [hunk 3]
```diff
@@ -2,74 +2,176 @@
...
     def test_create_vm_simple(self):
         """
         We can create a new VM using mostly default values.
+        """
+        xenserver = FakeXenServer()
+        self.setup_xs_storage(xenserver)
+        self.setup_xs_networks(xenserver)
+        session = xenserver.getSession()
+        template = self.setup_template("footempl")
+        vm = self.setup_vm("foovm", template)
```
Comment (JayH5) [inline xenserver/tests/test_tasks.py]: Should this stuff not be in a setup method or something?

File: xenserver/tests/test_tasks.py [hunk 4]
```diff
@@ -2,74 +2,176 @@
...
+        # _create_vm().
+        assert xenserver.VIFs.keys() == [ev.VIF.value]
+        assert ExpectedXenServerVIF(
+            device="0", VM=vm.xsref, network=self.pub_net,
+        ) == xenserver.VIFs[ev.VIF.value]
+
+        # The VM should be started.
+        assert xenserver.VM_operations == [(vm.xsref, "start")]
+
+    # @pytest.mark.xfail
```
Comment (JayH5) [inline xenserver/tests/test_tasks.py]: Can be removed?

File: xenserver/tests/test_views.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    def test_provision_simple(self, monkeypatch, settings, admin_client):
+        """
+        We can create a new VM using mostly default values.
+        """
+        createvm_calls = []
+        monkeypatch.setattr(tasks, 'getSession', lambda *a, **kw: "session")
+        monkeypatch.setattr(
+            tasks, '_create_vm', lambda *a: createvm_calls.append(a))
+        settings.CELERY_ALWAYS_EAGER = True
```
Comment (JayH5) [inline xenserver/tests/test_views.py]: Maybe a comment about this `monkeypatch` and `CELERY_ALWAYS_EAGER` stuff.. I don't really know what's going on there..

File: xenserver/tests/test_views.py (hunk 2 above)
Comment (jerith) [inline xenserver/tests/test_views.py]: Tyop, thanks.

File: xenserver/tests/test_tasks.py (hunk 3 above)
Comment (jerith) [inline xenserver/tests/test_tasks.py]: Some tests will want it to be different. I just don't have any of those tests yet.

File: xenserver/tests/test_tasks.py (hunk 4 above)
Comment (jerith) [inline xenserver/tests/test_tasks.py]: Quite.

File: xenserver/tests/test_views.py (hunk 5 above)
Comment (jerith) [inline xenserver/tests/test_views.py]: I'll factor this out and comment it.

**Comment by JayH5** [comment]
I don't fully understand the Xen stuff, but the code looks fine to me :+1:

**Comment by calston** [comment]
:+1:

**Pull request merged**